include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,606 @@
+// number of black nodes.
+
+template <class T, class V>
+class RBTree {
+  // Colors used by the red black tree: Red (R), Black (B), and Double Black
+  // (BB). BB is a transitory color that allows to temporarily preserve the
+  // black invariant when certain operations occur, and is gradually removed
+  // while they are being completed. A red-black tree is never expected to have
+  // any nodes or leaves of color BB after an operation is performed on it.
+  enum class Color { R, B, BB };
+
+  // Return the "previous" color, i.e., the one containing one less black.
+  static Color minus_one_color(Color c) {
+    switch (c) {
+    case Color::B: return Color::R;
+    case Color::BB: return Color::B;
+    default: CONSTRUCT_MSG_AND_THROW("Unexpected color");
+    }
+  }
+
+  // Virtual class, for a node of the red-black tree.
+  struct Node {
+    Node(Color c)
+        : c_(c)
+        , s_(0) { }
+    Color c_; // Color of this tree Node
+    size_t s_; // Size of the tree with root Node
+
+    // Return true if this object is a Leaf.
+    virtual bool is_leaf() const = 0;
+    virtual ~Node() = default;
+  };
+
+  // A Leaf of the red-black tree.
+  struct Leaf : public Node {
+    // Create a new Leaf object with the given color. Only B and BB colors are
+    // valid arguments.
+    Leaf(Color c)
+        : Node(c) {
+      assert(c == Color::B || c == Color::BB);
+    }
+    virtual bool is_leaf() const override { return true; }
+    virtual ~Leaf() = default;
+  };
+
+  // An internal node of the red-black tree.
+  struct InternalNode : public Node {
+    // Create a new InternalNode object with the given lft and rgt children,
+    // key and value, and color.
+    InternalNode(
+        Color c, std::shared_ptr<const Node> lft, T key, V val,
+        std::shared_ptr<const Node> rgt)
+        : Node(c)
+        , lft_(lft)
+        , key_(key)
+        , val_(val)
+        , rgt_(rgt) {
+      this->s_ = 1 + lft_->s_ + rgt_->s_;
+    }
+    std::shared_ptr<const Node> lft_; // Left child
+    T key_; // Node key
+    V val_; // Node value
+    std::shared_ptr<const Node> rgt_; // Right child
+
+    virtual bool is_leaf() const override { return false; }
+    virtual ~InternalNode() = default;
+  };
+
+  // Create an empty red-black tree, with the specified color. Only B and BB
+  // are valid colors for this constructor.
+  RBTree(Color c)
+      : root_(std::make_shared<const Leaf>(c)) { }
+
+  // Create a red-black tree, with a root of the spefified color, key and value,
+  // and children lft and rgt.
+  RBTree(Color c, RBTree const &lft, T key, V val, RBTree const &rgt)
+      : root_(std::make_shared<const InternalNode>(
+          c, lft.root_, key, val, rgt.root_)) {
+    assert(lft.is_empty() || lft.root_key() < key);
+    assert(rgt.is_empty() || key < rgt.root_key());
+  }
+
+  // Copy constructor.
+  explicit RBTree(std::shared_ptr<const Node> node)
+      : root_(node) { }
```

Comment:
Nitpicking but this isn't a true C++ copy constructor; it constructs a new tree from a _node_ rather than from a different tree.

Reply:
Addressed.